pkg/tfbridge/rawstate.go

```diff
@@ -0,0 +1,867 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tfbridge
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"math/big"
+	"os"
+
+	"github.com/hashicorp/go-cty/cty"
+	ctyjson "github.com/hashicorp/go-cty/cty/json"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/resource"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/cmdutil"
+	"github.com/pulumi/pulumi/sdk/v3/go/common/util/contract"
+
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfbridge/log"
+	shim "github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/pkg/tfshim/walk"
+	"github.com/pulumi/pulumi-terraform-bridge/v3/unstable/propertyvalue"
+)
+
+// rawStateDeltaKey is where [rawStateDelta] is stored under [metaKey].
+const rawStateDeltaKey = "__pulumi_raw_state_delta"
+
+// Represents an efficient delta between PropertyMap and cty.Value representations of resource state.
+//
+// Pulumi stores Terraform resource states as PropertyMap in the state file. But providers expect to interact with
+// cty.Value versions of the state. This is especially pertinent when upgrading provider versions. The new Terraform
+// provider code on V-next expects to receive the verbatim raw state JSON written by V-prev and possibly massage it
+// with state upgrade code toward the newly changed schema.
+//
+// cty.Value has slightly more information that strictly necessary but is taken as an approximation for the Terraform
+// raw state in this code.
+//
+// It is not possible to reconstruct cty.Value from PropertyMap because the transformation is schema aware, and the
+// V-prev schema is not available at read time.
+//
+// It should be possible to store cty.Value alongside PropertyMap in Pulumi state, but this approach would require
+// additional storage. More pertinently, it would complicate editing state files in scenarios requiring state repair.
+//
+// The code takes a hybrid approach instead. It computes a RawStateDelta representing the difference between a
+// PropertyMap and a cty.Value, so that cty.Value need not be stored itself. In typical scenarios RawStateDelta is
+// fairly small. It falls back to storing the entirety of cty.Value if an efficient encoding cannot be computed.
+//
+// At read time, PropertyMap and RawStateDelta are read, and the original cty.Value is reconstructed from the pair.
+type RawStateDelta struct {
+	TypedNull *typedNullDelta `json:"null,omitempty"`
+	Pluralize *pluralizeDelta `json:"plu,omitempty"`
+	Map       *mapDelta       `json:"map,omitempty"`
+	Obj       *objDelta       `json:"obj,omitempty"`
+	Array     *arrayDelta     `json:"arr,omitempty"`
+	Set       *setDelta       `json:"set,omitempty"`
+	Asset     *assetDelta     `json:"asset,omitempty"`
+	Num       *numDelta       `json:"num,omitempty"`
+	Replace   *replaceDelta   `json:"replace,omitempty"`
```

Comment:
I think this needs naming information so that if we introduce a name override we can correctly reconstruct the TF name from this representation.

Reply:
That's handled in Obj.